source/common/stats/thread_local_store.cc

```diff
@@ -204,29 +253,123 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
-  if (!central_ref) {
-    std::vector<Tag> tags;
-    std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
-    central_ref.reset(
-        new HistogramImpl(final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
-  }
+  std::vector<Tag> tags;
+  std::string tag_extracted_name = parent_.getTagsForName(name, tags);
+  TlsHistogramSharedPtr hist_tls_ptr(
+      new ThreadLocalHistogramImpl(name, std::move(tag_extracted_name), std::move(tags)));
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[name];
+  central_ref->addTlsHistogram(hist_tls_ptr);
 
   if (tls_ref) {
-    *tls_ref = central_ref;
+    *tls_ref = hist_tls_ptr;
   }
+  return *hist_tls_ptr;
+}
 
-  return *central_ref;
+ThreadLocalHistogramImpl::ThreadLocalHistogramImpl(const std::string& name,
+                                                   std::string&& tag_extracted_name,
+                                                   std::vector<Tag>&& tags)
+    : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), current_active_(0),
+      flags_(0), created_thread_id_(std::this_thread::get_id()) {
+  histograms_[0] = hist_alloc();
+  histograms_[1] = hist_alloc();
+}
+
+ThreadLocalHistogramImpl::~ThreadLocalHistogramImpl() {
+  hist_free(histograms_[0]);
+  hist_free(histograms_[1]);
+}
+
+void ThreadLocalHistogramImpl::recordValue(uint64_t value) {
+  ASSERT(std::this_thread::get_id() == created_thread_id_);
+  hist_insert_intscale(histograms_[current_active_], value, 0, 1);
+  flags_ |= Flags::Used;
+}
+
+void ThreadLocalHistogramImpl::merge(histogram_t* target) {
+  const uint64_t other_histogram_index = otherHistogramIndex();
+  hist_accumulate(target, &histograms_[other_histogram_index], 1);
+  hist_clear(histograms_[other_histogram_index]);
+}
+
+ParentHistogramImpl::ParentHistogramImpl(const std::string& name, Store& parent, TlsScope& tlsScope,
+                                         std::string&& tag_extracted_name, std::vector<Tag>&& tags)
+    : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent),
+      tlsScope_(tlsScope), interval_histogram_(hist_alloc()), cumulative_histogram_(hist_alloc()),
+      interval_statistics_(interval_histogram_), cumulative_statistics_(cumulative_histogram_) {}
+
+ParentHistogramImpl::~ParentHistogramImpl() {
+  hist_free(interval_histogram_);
+  hist_free(cumulative_histogram_);
+}
+
+void ParentHistogramImpl::recordValue(uint64_t value) {
+  Histogram& tls_histogram = tlsScope_.tlsHistogram(name());
+  tls_histogram.recordValue(value);
+  parent_.deliverHistogramToSinks(*this, value);
+}
+
+bool ParentHistogramImpl::used() const {
+  std::unique_lock<std::mutex> lock(merge_lock_);
+  return usedLockHeld();
+}
+
+void ParentHistogramImpl::merge() {
+  std::unique_lock<std::mutex> lock(merge_lock_);
+  if (usedLockHeld()) {
+    hist_clear(interval_histogram_);
+    for (const TlsHistogramSharedPtr& tls_histogram : tls_histograms_) {
```

Comment:
Matt commented earlier that you could capture the histogram-array and then unlock before doing the expensive merge.  I'm not actually sure what would contend with merge_lock_ though, so I don't know how critical it is in practice.  But a comment or TODO anyway would be good.

Reply:
I have added a comment explaining why it is not a big deal.